sql/sql_repl.cc

```diff
@@ -4058,17 +4058,15 @@ bool change_master(THD* thd, Master_info* mi, bool *master_info_added)
   if (lex_mi->port)
     mi->port = lex_mi->port;
   if (lex_mi->connect_retry)
-    mi->connect_retry = lex_mi->connect_retry;
+    lex_mi->connect_retry(mi);
   if (lex_mi->retry_count)
   {
-    mi->retry_count= lex_mi->retry_count;
+    lex_mi->retry_count(mi);
+    // also reset the counter in case `connects_tried > master_retry_count`
     mi->connects_tried= 0;
   }
-  if (lex_mi->heartbeat_opt != LEX_MASTER_INFO::LEX_MI_UNCHANGED)
-    mi->heartbeat_period = lex_mi->heartbeat_period;
-  else
-    mi->heartbeat_period= (float) MY_MIN(SLAVE_MAX_HEARTBEAT_PERIOD,
-                                      (slave_net_timeout/2.0));
+  if (lex_mi->heartbeat_period)
+    lex_mi->heartbeat_period(mi);
   mi->received_heartbeats= 0; // counter lives until master is CHANGEd
```

Comment:
What was the intent to have CHANGE MASTER always reset this?
Was it because `master_heartbeat_period` is always changed regardless of whether it was configured in the CHANGE MASTER?
* (c.f. #3974)